File: AtCoder/ABC/165/python/C_Many_Requirements.py
N, M, Q = map(int, input().split())
X = [list(map(int, input().split())) for _ in range(Q)]
A = [1] * N


def dfs(i):
    res = 0
    if i == N:
        for a, b, c, d in X:
            if A[b - 1] - A[a - 1] == c:
                res += d
        return res
    for j in range(A[i - 1], M + 1):
        A[i] = j
        res = max(res, dfs(i + 1))
    return res


print(dfs(1))

File: AtCoder/ABC/165/cpp/C_Many_Requirements.cpp
#include <bits/stdc++.h>
#define MAX_N 10
#define MAX_Q 50
using namespace std;

int N, M, Q;
int a[MAX_Q], b[MAX_Q], c[MAX_Q], d[MAX_Q];
int A[MAX_N];

int dfs(int i) {
  int res = 0;
  if (i == N) {
    for (int j = 0; j < Q; ++j) {
      if (A[b[j]] - A[a[j]] == c[j]) {
        res += d[j];
      }
    }
    return res;
  }
  for (int j = A[i - 1]; j <= M; ++j) {
    A[i] = j;
    res = max(res, dfs(i + 1));
  }
  return res;
}

void solve() {
  A[0] = 1;
  printf("%d\n", dfs(1));
}

int main() {
  cin >> N >> M >> Q;
  for (int i = 0; i < Q; ++i) {
    cin >> a[i] >> b[i] >> c[i] >> d[i];
    --a[i];
    --b[i];
  }
  solve();
}